### PR TITLE
Refactor context package

### PR DESCRIPTION
**Pull request by akutz**
/kind feature

**What this PR does / why we need it**:
This patch refactors the context package to provide additional functionality such as:

* Distinct contexts for a controller manager and controller
* A manager package for running parallel managers in the same process
* A distinct session package for getting and creating vSphere sessions

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by akutz** [comment]
Hi @figo / @frapposelli,

Could you please review this when you have a moment?

File: controllers/vspherecluster_controller.go [hunk 1]
```diff
@@ -17,106 +17,144 @@ limitations under the License.
...
+		Name:                     controllerNameShort,
+		Recorder:                 record.New(mgr.GetEventRecorderFor(controllerNameLong)),
+		Logger:                   ctx.Logger.WithName(controllerNameShort),
+	}
 
-	logger := r.Log.WithName(controllerName).
-		WithName(fmt.Sprintf("namespace=%s", req.Namespace)).
-		WithName(fmt.Sprintf("vsphereCluster=%s", req.Name))
+	controlledType := &infrav1.VSphereCluster{}
+	controlledTypeName := reflect.TypeOf(controlledType).Elem().Name()
```
Comment (frapposelli) [inline controllers/vspherecluster_controller.go]: How hot is this code path? from a quick glance looks like it's only called when attaching a new cluster to a manager, but just want to be sure 😄

File: controllers/vspheremachine_controller.go [hunk 2]
```diff
@@ -17,150 +17,181 @@ limitations under the License.
...
+		Recorder:                 record.New(mgr.GetEventRecorderFor(controllerNameLong)),
+		Logger:                   ctx.Logger.WithName(controllerNameShort),
+	}
 
-	logger := r.Log.
-		WithName(controllerName).
-		WithName(fmt.Sprintf("namespace=%s", req.Namespace)).
-		WithName(fmt.Sprintf("vsphereMachine=%s", req.Name))
+	controlledType := &infrav1.VSphereMachine{}
+	controlledTypeName := reflect.TypeOf(controlledType).Elem().Name()
```
Comment (frapposelli) [inline controllers/vspheremachine_controller.go]: Same comment as above

File: pkg/manager/manager.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	bootstrapv1 "sigs.k8s.io/cluster-api-bootstrap-provider-kubeadm/api/v1alpha2"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha2"
+	ctrlmgr "sigs.k8s.io/controller-runtime/pkg/manager"
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-vsphere/api/v1alpha2"
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/context"
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/record"
+)
+
+// Manager is a GCM controller manager.
```
Comment (frapposelli) [inline pkg/manager/manager.go]: s/GCM/CAPV?

File: pkg/manager/manager.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+type Manager interface {
+	ctrlmgr.Manager
+
+	// GetContext returns the controller manager's context.
+	GetContext() *context.ControllerManagerContext
+}
+
+// New returns a new GCM controller manager.
```
Comment (frapposelli) [inline pkg/manager/manager.go]: s/GCM/CAPV?

File: pkg/manager/options.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/context"
+)
+
+// AddToManagerFunc is a function that can be optionally specified with
+// the manager's Options in order to explicitly decide what controllers and
+// webhooks to add to the manager.
+type AddToManagerFunc func(*context.ControllerManagerContext, ctrlmgr.Manager) error
+
+// Options describes the options used to create a new GCM manager.
```
Comment (frapposelli) [inline pkg/manager/options.go]: s/GCM/CAPV?

File: controllers/vspherecluster_controller.go (hunk 1 above)
Comment (akutz) [reply controllers/vspherecluster_controller.go]: This is only called when attaching the controller to a new controller manager.

File: controllers/vspheremachine_controller.go (hunk 2 above)
Comment (akutz) [reply controllers/vspheremachine_controller.go]: This is only called when attaching the controller to a new controller manager.

**Comment by frapposelli** [comment]
/hold cancel
/lgtm

👍🏻

**Pull request merged**